File: tools/_validate_tr_strings.py
#!/usr/bin/env python3
"""Compare EN vs TR string resources: names, dupes, placeholder multisets."""
from __future__ import annotations

import re
import sys
from collections import Counter
from pathlib import Path

ROOT = Path(__file__).resolve().parents[1]
EN = ROOT / "app" / "src" / "main" / "res" / "values" / "strings.xml"
TR = ROOT / "app" / "src" / "main" / "res" / "values-tr" / "strings.xml"

STRING_RE = re.compile(
    r'<string\s+name="([^"]+)"([^>]*)>(.*?)</string>', re.DOTALL
)


def parse(path: Path) -> dict[str, str]:
    text = path.read_text(encoding="utf-8")
    items = STRING_RE.findall(text)
    names = [n for n, _, _ in items]
    dupes = [n for n, c in Counter(names).items() if c > 1]
    if dupes:
        raise SystemExit(f"Duplicate keys in {path}: {dupes}")
    return {n: body for n, _, body in items}


def extract_placeholders(s: str) -> list[str]:
    """Extract conversion specs: %%, %s, %d, %1$d, %2$s, etc."""
    tokens: list[str] = []
    i = 0
    while i < len(s):
        if s[i] != "%":
            i += 1
            continue
        if i + 1 < len(s) and s[i + 1] == "%":
            tokens.append("%%")
            i += 2
            continue
        m = re.match(r"%(?:(\d+)\$)?([sdif])", s[i:])
        if m:
            pos, typ = m.group(1), m.group(2)
            tokens.append(f"%{pos + '$' if pos else ''}{typ}")
            i += m.end()
            continue
        i += 1
    return tokens


def normalize_multiset(tokens: list[str]) -> Counter:
    """Normalize %d and %1$d carefully into typed multiset."""
    c: Counter = Counter()
    for t in tokens:
        if t == "%%":
            c["%%"] += 1
            continue
        m = re.fullmatch(r"%(?:(\d+)\$)?([sdif])", t)
        if not m:
            c[t] += 1
            continue
        pos, typ = m.group(1), m.group(2)
        # Keep positional identity when present; else type-only bucket
        key = f"%{pos}${typ}" if pos else f"%{typ}"
        c[key] += 1
    return c


def main() -> int:
    en = parse(EN)
    tr = parse(TR)
    en_names = set(en)
    tr_names = set(tr)

    missing = sorted(en_names - tr_names)
    extra = sorted(tr_names - en_names)
    ph_mismatches = []

    for name in sorted(en_names & tr_names):
        en_ph = extract_placeholders(en[name])
        tr_ph = extract_placeholders(tr[name])
        # Sequence of conversion types must match (order + types)
        en_types = [
            "%%" if t == "%%" else re.fullmatch(r"%(?:\d+\$)?([sdif])", t).group(1)
            for t in en_ph
        ]
        tr_types = [
            "%%" if t == "%%" else re.fullmatch(r"%(?:\d+\$)?([sdif])", t).group(1)
            for t in tr_ph
        ]
        if en_types != tr_types or normalize_multiset(en_ph) != normalize_multiset(
            tr_ph
        ):
            # Allow positional reordering only when both use explicit positionals
            # and multisets of pos+type match; still require same type multiset
            if normalize_multiset(en_ph) != normalize_multiset(tr_ph) or Counter(
                en_types
            ) != Counter(tr_types):
                ph_mismatches.append(
                    {
                        "name": name,
                        "en": en_ph,
                        "tr": tr_ph,
                        "en_body": en[name],
                        "tr_body": tr[name],
                    }
                )

    header = TR.read_text(encoding="utf-8")[:80]
    header_ok = TR.read_text(encoding="utf-8").startswith(
        '<?xml version="1.0" encoding="utf-8"?>\n<resources>'
    )

    print(f"EN strings: {len(en)}")
    print(f"TR strings: {len(tr)}")
    print(f"Missing in TR: {missing}")
    print(f"Extra in TR: {extra}")
    print(f"Placeholder mismatches: {len(ph_mismatches)}")
    for m in ph_mismatches[:20]:
        print(f"  - {m['name']}: EN={m['en']} TR={m['tr']}")
    print(f"Header OK: {header_ok}")
    print(f"Header preview: {header!r}")

    ok = (
        len(en) == len(tr)
        and not missing
        and not extra
        and not ph_mismatches
        and header_ok
    )
    print("RESULT:", "PASS" if ok else "FAIL")
    return 0 if ok else 1


if __name__ == "__main__":
    sys.exit(main())

File: tools/_validate_tr_strings.mjs
/**
 * EN vs TR validation (Node equivalent of _validate_tr_strings.py)
 * - same set of string names
 * - no duplicate keys in TR
 * - same placeholder token multiset per key
 */
import fs from "fs";
import path from "path";
import { fileURLToPath } from "url";

const ROOT = path.resolve(path.dirname(fileURLToPath(import.meta.url)), "..");
const EN = path.join(ROOT, "app/src/main/res/values/strings.xml");
const TR = path.join(ROOT, "app/src/main/res/values-tr/strings.xml");

const STRING_RE = /<string\s+name="([^"]+)"([^>]*)>([\s\S]*?)<\/string>/g;

function parse(xmlPath) {
  const text = fs.readFileSync(xmlPath, "utf8");
  const names = [];
  const map = {};
  let m;
  const re = new RegExp(STRING_RE.source, "g");
  while ((m = re.exec(text))) {
    names.push(m[1]);
    map[m[1]] = m[3];
  }
  const seen = new Set();
  const dupes = [];
  for (const n of names) {
    if (seen.has(n)) dupes.push(n);
    seen.add(n);
  }
  return { text, names, map, dupes };
}

function extractPlaceholders(s) {
  const tokens = [];
  let i = 0;
  while (i < s.length) {
    if (s[i] !== "%") {
      i++;
      continue;
    }
    if (s[i + 1] === "%") {
      tokens.push("%%");
      i += 2;
      continue;
    }
    const m = /^%(?:(\d+)\$)?([sdif])/.exec(s.slice(i));
    if (m) {
      tokens.push(m[0]);
      i += m[0].length;
      continue;
    }
    i++;
  }
  return tokens;
}

function normalizeMultiset(tokens) {
  const c = new Map();
  for (const t of tokens) {
    if (t === "%%") {
      c.set("%%", (c.get("%%") || 0) + 1);
      continue;
    }
    const m = /^%(?:(\d+)\$)?([sdif])$/.exec(t);
    if (!m) {
      c.set(t, (c.get(t) || 0) + 1);
      continue;
    }
    const pos = m[1];
    const typ = m[2];
    // Normalize: when comparing EN %d with TR %d → "%d"
    // When either uses positional, keep "%N$type" so reordering is detectably intentional
    const key = pos ? `%${pos}$${typ}` : `%${typ}`;
    c.set(key, (c.get(key) || 0) + 1);
  }
  return [...c.entries()].sort((a, b) => a[0].localeCompare(b[0]));
}

function typeSequence(tokens) {
  return tokens.map((t) => {
    if (t === "%%") return "%%";
    return /^%(?:\d+\$)?([sdif])$/.exec(t)[1];
  });
}

const en = parse(EN);
const tr = parse(TR);

const enSet = new Set(en.names);
const trSet = new Set(tr.names);
const missing = [...enSet].filter((n) => !trSet.has(n)).sort();
const extra = [...trSet].filter((n) => !enSet.has(n)).sort();

const phMismatches = [];
for (const name of [...enSet].sort()) {
  if (!tr.map[name]) continue;
  const ep = extractPlaceholders(en.map[name]);
  const tp = extractPlaceholders(tr.map[name]);
  const enTypes = typeSequence(ep);
  const trTypes = typeSequence(tp);
  const enMs = JSON.stringify(normalizeMultiset(ep));
  const trMs = JSON.stringify(normalizeMultiset(tp));

  // Require same ordered type sequence (placeholders stay aligned),
  // and same normalized multiset.
  // If EN uses positionals and TR reorders, multiset of pos$key must still match.
  const enHasPos = ep.some((t) => /%\d+\$/.test(t));
  const trHasPos = tp.some((t) => /%\d+\$/.test(t));
  let ok = enTypes.join() === trTypes.join();
  if (enHasPos || trHasPos) {
    ok = enMs === trMs;
  } else {
    ok = ok && enMs === trMs;
  }
  if (!ok) {
    phMismatches.push({ name, en: ep, tr: tp, enBody: en.map[name], trBody: tr.map[name] });
  }
}

const headerOk = tr.text.startsWith(
  '<?xml version="1.0" encoding="utf-8"?>\n<resources>'
);

const report = {
  enCount: en.names.length,
  trCount: tr.names.length,
  trUnique: trSet.size,
  trDupes: tr.dupes,
  missing,
  extra,
  phMismatchCount: phMismatches.length,
  phMismatches,
  headerOk,
  result:
    en.names.length === tr.names.length &&
    tr.dupes.length === 0 &&
    missing.length === 0 &&
    extra.length === 0 &&
    phMismatches.length === 0 &&
    headerOk
      ? "PASS"
      : "FAIL",
};

console.log(JSON.stringify(report, null, 2));
process.exit(report.result === "PASS" ? 0 : 1);
